test/E2E.test.ts

```diff
@@ -0,0 +1,473 @@
+    });
+
+    it("should have executed job successfully", async function () {
+      await keeperRegistryPerformUpkeep(
+        registry,
+        registrySigners,
+        upkeepId,
+        KeeperRegistryParams.F
+      );
+
+      const [workable, reasonBytes] = await job.workable(
+        formatBytes32String("test")
+      );
+      const reason = parseBytes32String(reasonBytes.padEnd(66, "0"));
+      expect(reason).to.equal("Timer hasn't elapsed");
+      expect(workable).to.equal(false);
+    });
+  });
+
+  describe("Refund upkeep", function () {
+    let pool: IUniswapV3Pool;
+    let linkTokensLiquidity: BigNumber;
+    let paymentTokensLiquidity: BigNumber;
+
+    before(async function () {
+      linkTokensLiquidity = ethers.utils.parseEther("100.0");
+      paymentTokensLiquidity = ethers.utils.parseEther("100.0");
+      pool = await setupPool(
+        linkToken,
+        paymentToken,
+        owner,
+        linkTokensLiquidity,
+        paymentTokensLiquidity
+      );
+    });
+
+    beforeEach(async function () {
+      await network.provider.send("evm_increaseTime", [
+        VestParams.DURATION_IN_SEC_VESTING_PLAN,
+      ]);
+      await network.provider.send("evm_mine");
+
+      expect(
+        await pool.liquidity(),
+        "should have liquidity in link/payment token pool"
+      ).to.be.gt(0);
+
+      expect(
+        await dssVest.unpaid(vestId),
+        "should have accrued tokens in vest"
+      ).to.equal(VestParams.CAP);
+    });
+
+    it("should refund upkeep when balance falls below threshold", async function () {
+      const { balance: preFundBalance } = await registry
+        .connect(ethers.constants.AddressZero)
+        .callStatic.getUpkeep(upkeepId);
+
+      // increase threshold above balance so it has to refund upkeep
+      await topUp.setThreshold(preFundBalance.add(1));
+
+      await keeperRegistryPerformUpkeep(
+        registry,
+        registrySigners,
+        upkeepId,
+        KeeperRegistryParams.F
+      );
+      const { balance: postFundBalance } = await registry
+        .connect(ethers.constants.AddressZero)
+        .callStatic.getUpkeep(upkeepId);
+
+      expect(postFundBalance).to.be.gt(preFundBalance);
+
+      const linkUpkeepRecieved = postFundBalance.sub(preFundBalance);
+      const paymentTransferedToDaiJoin = linkTokensLiquidity.sub(
+        TopUpParams.MAX_DEPOSIT_AMT_PAYMENT_TOKEN
+      );
+
+      console.log(paymentTransferedToDaiJoin);
+      console.log("dai join address:", daiJoinMock.address);
+      const vowBalance = await paymentToken.balanceOf(VOW);
+
+      // tokens recieved should be between MAX_DEPOSIT_AMT_PAYMENT_TOKEN and MAX_DEPOSIT_AMT_PAYMENT_TOKEN/2
+      expect(linkUpkeepRecieved).to.be.lte(
+        TopUpParams.MAX_DEPOSIT_AMT_PAYMENT_TOKEN
+      );
+      expect(linkUpkeepRecieved).to.be.gte(
+        TopUpParams.MAX_DEPOSIT_AMT_PAYMENT_TOKEN.div(2)
+      );
+
+      expect(vowBalance).to.eq(paymentTransferedToDaiJoin);
```

Comment:
This is testing `dss-vest` functionality, which is out of the scope. 

The `linkUpkeepReceived` should be the expected tokens out of the swap.